src/primitives.rs

```diff
@@ -0,0 +1,161 @@
+use super::requests;
+
+use crate::requests::RequestApiVersions;
+use bytes::Buf;
+use core::borrow::{Borrow, BorrowMut};
+use std::fmt::{Error, Formatter};
+use DecodeErrors::BadData;
+
+enum ApiKeys {
+    Metadata = 3,
+    ApiVersions = 18,
+}
+
+pub enum ApiVersions {
+    Version0,
+    Version1,
+    Version2,
+}
+
+#[derive(Debug)]
+pub enum DecodeErrors {
+    BufferUnderflow,
+    BufferOverflow,
+    BadData,
+}
+
+pub trait Message {
+    fn get_size(&self) -> i32;
```

Comment:
Nit: typically in Rust, it's more idiomatic for accessor/"getter" methods to just be named `property`, rather than `get_property`, so I'd call this
```suggestion
    fn size(&self) -> i32;
```

Reply:
Also, is the message size being signed something specified in Kafka? Because naively, it seems very strange for this to be an `i32` IMO --- what is the meaning of a message with size -20?

Reply:
`i32` is what the protocol says at https://kafka.apache.org/protocol.html#protocol_common . In fact, Kafka supports `u32` (`UINT32`) but it is never used in the protocol page. Weird to me.

Reply:
Yeah, that definitely seems surprising. I suppose if it's what the protocol specifies, `i32` is correct.